### PR TITLE
[Merged by Bors] - feat: the empty line in commands linter

### PR DESCRIPTION
**Pull request by adomani**
This linter flags empty lines within a command.

It allows empty lines within doc-strings, module-docs and a couple of other "sensible" places.

It also skips files that are likely to contain meta-code, since there the use of empty lines in definition is more widespread.

This PR continues the work from #25236.

### CODE REVIEW

**Comment by adomani** [comment]
## Comments from Original PR #25236

*This section contains 8 comment(s) from the original PR, excluding bot comments.*

---

**@github-actions** ([2025-05-27 11:59 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2912246700)):
### PR summary [ac10029deb](https://github.com/leanprover-community/mathlib4/pull/25236/commits/b5d3fd9dd810fe00a08883eb59b00a77b70f6015)

Import changes exceeding 2%

| %      | File |
| -      | -    |
| +3.57% | `Mathlib.Init` |

#### Import changes for modified files

No significant changes to the import graph


<details><summary>Import changes for all files</summary>

|Files|Import difference|
|-|-|
There are 6477 files with changed transitive imports taking up over 272804 characters: this is too many to display!
You can run `scripts/import_trans_difference.sh all` locally to see the whole output.

</details>

---

<details><summary>

#### Declarations diff

</summary>

`+ AllowEmptyLines`
`+ F`
`+ F₀`
`+ Lean.Substring.getRange`
`+ SkippedFileSegments`
`+ WithAString`
`+ emptyLineLinter`
`+ filter`
`+ filterMap`
`+ filterMapM`
<details>
  <summary>You can run this locally as follows</summary>


```bash
## summary with just the declaration names:
./scripts/declarations_diff.sh <optional_commit>

## more verbose report:
./scripts/declarations_diff.sh long <optional_commit>
```
</details>

The doc-module for `script/declarations_diff.sh` contains some details about this script.

</details>

---

<details><summary>No changes to technical debt.</summary>

You can run this locally as
```
./scripts/technical-debt-metrics.sh pr_summary
```
* The `relative` value is the weighted *sum* of the differences with weight given by the *inverse* of the current value of the statistic.
* The `absolute` value is the `relative` value divided by the total sum of the inverses of the current values (i.e. the weighted *average* of the differences).
</details>

---

**@bryangingechen** ([2025-05-28 17:21 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2917076670)):
Could you (or someone else) open a PR to the style guide also so we can have some guidance for contributors in place before this is enforced?

---

**@adomani** ([2025-05-28 20:03 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2917473174)):
@bryangingechen thank you for the suggestion!  [community#637](https://github.com/leanprover-community/leanprover-community.github.io/pull/637)

---

**@leanprover-community-bot-assistant** ([2025-06-03 11:10 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2934728335)):
This pull request has conflicts, please merge `master` and resolve them.

---

**@grunweg** ([2025-06-03 12:53 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2935101589)):
The tests also looks nice and thorough. Probably, there may be edge cases that we'll discover, but lots of them are already tested - this looks safe to try.

---

**@grunweg** ([2025-06-03 12:57 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2935116716)):
So, apart from the one question I asked above, this looks ready for me.

---

**@grunweg** ([2025-06-03 13:30 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2935252994)):
Thanks!
maintainer merge

---

**@github-actions** ([2025-06-03 13:30 UTC](https://github.com/leanprover-community/mathlib4/pull/25236#issuecomment-2935254843)):
🚀 Pull request has been placed on the maintainer queue by grunweg.

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Mathlib.Tactic.Linter.Header
+
+/-!
+#  The "emptyLine" linter
```
Comment (bryangingechen) [inline Mathlib/Tactic/Linter/EmptyLine.lean]:
```suggestion
# The "emptyLine" linter
```

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+end Lean.Syntax
+
+namespace Mathlib.Linter
+
+/--
+The "emptyLine" linter emits a warning on empty lines inside a command, but outside of a
+doc-string/module-doc.
+
+The linter is only active when there are no other warning, so as to not add noise when developing
```
Comment (bryangingechen) [inline Mathlib/Tactic/Linter/EmptyLine.lean]:
```suggestion
The linter is only active when there are no other warnings, so as to not add noise when developing
```

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+#  The "emptyLine" linter
+
+The "emptyLine" linter emits a warning on empty lines inside a command, but outside of a
+doc-string/module-doc.
+-/
+
+open Lean Elab Linter
+
+/-- Retrieve the `String.Range` of a `Substring`. -/
+def Lean.Substring.getRange : Substring → String.Range
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: Why is this in the `Lean` namespace; isn't `Substring` in the root namespace?

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    | none => return nargs
+
+/--
+`filterMap stx f` takes as input a `Syntax` `stx` and a function `f : Syntax → α`.
+It produces the array of the `some` values that `f` takes while traversing `stx`.
+
+See `filterMapM` for a non-monadic version.
+-/
+def filterMap {α} (stx : Syntax) (f : Syntax → Option α) : Array α :=
+  stx.filterMapM (m := Id) f
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: Strictly speaking this should be
```suggestion
  Id.run <| stx.filterMapM fun x => pure (f x)
```

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+@[inherit_doc Mathlib.Linter.linter.style.emptyLine]
+def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.emptyLine (← getLinterOptions) do
+    return
+  -- The linter does not report anything on incomplete proofs, e.g. proofs containing `sorry`
+  -- or `stop`.
+  if (← get).messages.reportedPlusUnreported.any (!·.severity matches .information) then
+    return
+  if !((← getMainModule).components.filter SkippedFileSegments.contains).isEmpty then
+    return
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean:103]: This comment should be on the second if, right?

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+@[inherit_doc Mathlib.Linter.linter.style.emptyLine]
+def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.emptyLine (← getLinterOptions) do
+    return
+  -- The linter does not report anything on incomplete proofs, e.g. proofs containing `sorry`
+  -- or `stop`.
+  if (← get).messages.reportedPlusUnreported.any (!·.severity matches .information) then
+    return
+  if !((← getMainModule).components.filter SkippedFileSegments.contains).isEmpty then
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: `find?` will be faster than `filter` here

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    return
+  -- We ignore empty lines "after" the command finished.
+  let stx := stx.unsetTrailing
+  let allowed := stx.filter (AllowEmptyLines.contains ·.getKind)
+  let allowedRanges := allowed.filterMap (·.getRange?)
+  if let some str := stx.getSubstring? then
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: You can save yourself a level of indentation here with
```suggestion
  let some str := stx.getSubstring? | return
```
You should also do this before you do the possibly-expensive filtering.

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    if let one::rest := str.toString.trimRight.splitOn "\n\n" then
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: This one can also become an early return if this match fails.

Also this will fail your whitespace linter :)

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        if let some str := s.getTrailing?
+        then
+          -- Handle `where` and `where` fields.
+          if s.getAtomVal == "where" ||
+             s.isOfKind ``Parser.Term.structInstField ||
+             s.isOfKind ``Parser.Command.structSimpleBinder then
+            s.getTrailing?.map (·.getRange)
+          else
+            let strim := str.trim
+            if (strim.splitOn "\n\n").length != 1 then
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: There's no need to do a full split if you only care whether there are any matches; a find would be faster.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 3 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: You are right: I removed `Lean`!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 1 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: Done!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 4 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: Switched to `Id.run`!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: Fixed!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 6 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: I used `find?`, thanks!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 5 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean:103]: I think that the comment about incomplete proofs is where it should be: if there are any messages, then do nothing.

Maybe this will be clearer once I addressed the remaining comments!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 7 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: I used `return`, unindented, and moved this check as soon as `stx` has been defined with the correct meaning.

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 9 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: With the new `Slice`, there is a `contains` that I used.  Let me know if there is a way to avoid converting the substring to a string first, though!

File: Mathlib/Tactic/Linter/EmptyLine.lean (hunk 8 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: Thanks!

I combined the match also with the `rest.isEmpty` into a single `let ... | return`.

**Review by adomani — COMMENTED** [review]
I think that I incorporated all the suggestions: thank you very much!

File: Mathlib/Data/UInt.lean [hunk 10]
```diff
@@ -28,6 +28,9 @@ cautious here.
 
 @[expose] public section
 
+@[expose] public section
+
```
Comment (adomani) [inline Mathlib/Data/UInt.lean]:
```suggestion
```
this looks like a weird git resolution.

File: Mathlib/Tactic/Linter/EmptyLine.lean [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+  let allowedRanges := trails.insertMany allowedRanges
+  for (rg, before, after) in ranges do
+    if allowedRanges.any fun okRg ↦ okRg.start ≤ rg.start && rg.stop ≤ okRg.stop then
+      continue
+    -- `s` is a string of as many spaces (` `) as the characters of the previous line.
+    -- This, followed by the downarrow (`↓`) creates a pointer to an offending line break.
+    let s : String := .join <| List.replicate (before.length + 1) " "
+    Linter.logLint linter.style.emptyLine (.ofRange rg)
+      m!"Please, write a comment here or remove this line, \
+        but do not place empty lines within commands!\nContext:\n\
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/EmptyLine.lean]: Looking at the test output, I suggest to remove this newline: it appears like two blank lines (even though the second one has just the downward arrow).
```suggestion
        but do not place empty lines within commands!\nContext:\
```

Comment (adomani) [reply Mathlib/Tactic/Linter/EmptyLine.lean]: Thanks!  I accepted this suggestion and fixed the `#guard_msgs` in the test file.

**Comment by adomani** [comment]
Why did building mathlib take over 40mins in [this CI run](https://github.com/leanprover-community/mathlib4/actions/runs/19564015961/job/56021813880?pr=25945), where all I did in the previous one was edit a leaf file?

EDIT: Probably the reason is that [the previous run](https://github.com/leanprover-community/mathlib4/actions/runs/19553923521/job/55992221025?pr=25945) was unable to upload the cache to Azure, even though it had built mathlib and only failed due to the linting errors.

I am still a bit puzzled, since I think that I did download the cache to try and edit the file...
